client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/Role.java

```diff
@@ -0,0 +1,271 @@
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public final class Role implements ToXContentObject {
+
+    public static final ParseField CLUSTER = new ParseField("cluster");
+    public static final ParseField GLOBAL = new ParseField("global");
+    public static final ParseField INDICES = new ParseField("indices");
+    public static final ParseField APPLICATIONS = new ParseField("applications");
+    public static final ParseField RUN_AS = new ParseField("run_as");
+    public static final ParseField METADATA = new ParseField("metadata");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<Role, Void> PARSER = new ConstructingObjectParser<>("role_descriptor", false,
+            constructorObjects -> {
+                int i = 0;
+                final Collection<ClusterPrivilege> clusterPrivileges = (Collection<ClusterPrivilege>) constructorObjects[i++];
+                final ManageApplicationsPrivilege manageApplicationPrivileges = (ManageApplicationsPrivilege) constructorObjects[i++];
+                final Collection<IndicesPrivileges> indicesPrivileges = (Collection<IndicesPrivileges>) constructorObjects[i++];
+                final Collection<ApplicationResourcePrivileges> applicationResourcePrivileges =
+                        (Collection<ApplicationResourcePrivileges>) constructorObjects[i++];
+                final Collection<String> runAs = (Collection<String>) constructorObjects[i++];
+                final Map<String, Object> metadata = (Map<String, Object>) constructorObjects[i];
+                return new Role(clusterPrivileges, manageApplicationPrivileges, indicesPrivileges, applicationResourcePrivileges,
+                        runAs, metadata);
+            });
+
+    static {
+        PARSER.declareFieldArray(optionalConstructorArg(), (p, c) -> ClusterPrivilege.fromString(p.text()), CLUSTER,
+                ValueType.STRING_ARRAY);
+        PARSER.declareObject(optionalConstructorArg(), ManageApplicationsPrivilege.PARSER, GLOBAL);
+        PARSER.declareFieldArray(optionalConstructorArg(), IndicesPrivileges.PARSER, INDICES, ValueType.OBJECT_ARRAY);
+        PARSER.declareFieldArray(optionalConstructorArg(), ApplicationResourcePrivileges.PARSER, APPLICATIONS, ValueType.OBJECT_ARRAY);
+        PARSER.declareStringArray(optionalConstructorArg(), RUN_AS);
+        PARSER.<Map<String, Object>>declareObject(constructorArg(), (parser, c) -> parser.map(), METADATA);
+    }
+
+    private final Set<ClusterPrivilege> clusterPrivileges;
+    private final @Nullable ManageApplicationsPrivilege manageApplicationPrivileges;
+    private final Set<IndicesPrivileges> indicesPrivileges;
+    private final Set<ApplicationResourcePrivileges> applicationResourcePrivileges;
+    private final Set<String> runAs;
+    private final Map<String, Object> metadata;
```

Comment:
We don't have a `name`, server does not parse it 😞 . API will have to provide the name as a separate argument.

Reply:
I think that's a problem. I wonder if we should just make the server return a name?
We'll need to do something - I think a role really needs to know its name,

Reply:
I think it lends to the same strategy as with https://github.com/elastic/elasticsearch/blob/7e144b318ac34433ea971f2574723c2636c90e1b/client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/User.java#L37

Don't implement `ToXContentObject` by the Role object, and have a name field. The Request and Response objects will take care of the serialization and deserialization, and the client would use a Role with a name.

I'll do that in the follow-up PR also.